gaming/src/golang/profile-service/models/player.go

```diff
@@ -0,0 +1,211 @@
+	Password        string `json:"password" validate:"required_with=Player_name Email"` // not stored in DB
+	Password_hash   []byte `json:"password_hash"`                                       // stored in DB
+	created         time.Time
+	updated         time.Time
+	Stats           spanner.NullJSON `json:"stats"`
+	Account_balance big.Rat          `json:"account_balance"`
+	last_login      time.Time
+	is_logged_in    bool
+	valid_email     bool
+	Current_game    string `json:"current_game" validate:"omitempty,uuid4"`
+}
+
+func init() {
+	validate = validator.New()
+}
+
+// Helper function to read rows from Spanner.
+func readRows(iter *spanner.RowIterator) ([]spanner.Row, error) {
+	var rows []spanner.Row
+	defer iter.Stop()
+
+	for {
+		row, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		rows = append(rows, *row)
+	}
+
+	return rows, nil
+}
+
+// TODO complexity validation
+func hashPassword(pwd string) ([]byte, error) {
+	hash, err := bcrypt.GenerateFromPassword([]byte(pwd), bcrypt.DefaultCost)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return hash, nil
+}
+
+func validatePassword(pwd string, hash []byte) error {
+	return bcrypt.CompareHashAndPassword(hash, []byte(pwd))
+}
+
+func generateUUID() string {
+	return uuid.NewString()
+}
+
+func (p *Player) Validate() error {
+	validate = validator.New()
+	err := validate.Struct(p)
+	if err != nil {
+		return err
+	}
+
+	// validationErrors := err.(validator.ValidationErrors)
+	if _, ok := err.(*validator.InvalidValidationError); ok {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Player) AddPlayer(ctx context.Context, client spanner.Client) error {
+	// Validate based on struct validation rules
+	err := p.Validate()
+	if err != nil {
+		return err
+	}
+
+	// take supplied password+salt, hash. Store in user_password
+	passHash, err := hashPassword(p.Password)
+
+	if err != nil {
+		return errors.New("Unable to hash password")
+	}
+
+	p.Password_hash = passHash
+
+	// Generate UUIDv4
+	p.PlayerUUID = generateUUID()
+
+	// Initialize player stats
+	emptyStats := spanner.NullJSON{Value: PlayerStats{
+		Games_played: spanner.NullInt64{Int64: 0, Valid: true},
+		Games_won:    spanner.NullInt64{Int64: 0, Valid: true},
+	}, Valid: true}
+
+	// insert into spanner
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		stmt := spanner.Statement{
+			SQL: `INSERT players (playerUUID, player_name, email, password_hash, created, stats) VALUES
+					(@playerUUID, @playerName, @email, @passwordHash, CURRENT_TIMESTAMP(), @pStats)
+			`,
+			Params: map[string]interface{}{
+				"playerUUID":   p.PlayerUUID,
+				"playerName":   p.Player_name,
+				"email":        p.Email,
+				"passwordHash": p.Password_hash,
+				"pStats":       emptyStats,
+			},
+		}
+
+		_, err := txn.Update(ctx, stmt)
+		return err
+	})
+
+	// todo: Handle 'AlreadyExists' errors
```

Comment:
This could be avoided/handled by using an InsertOrUpdate mutation instead of an `INSERT` DML statement.

Reply:
Right, but I don't necessarily want to update if it already exists. More likely I want to throw the error, and perhaps a more specific error than if validation didn't succeed, and it would then be up to the frontend to handle the error(s).

Reply:
Ack